kiwi/system/kernel.py

```diff
@@ -90,6 +90,27 @@ def get_kernel(
                             filename=kernel_file,
                             version=version
                         )
+
+        boot_ostree_dir = os.sep.join([self.root_dir, 'boot/ostree'])
+        kernel_ostree_pattern = '.*/boot/ostree/.*/vmlinuz-(.*)'
+        if os.path.isdir(boot_ostree_dir):
+            for deployment in sorted(os.listdir(boot_ostree_dir)):
+                deployment_dir = os.sep.join([self.root_dir, 'boot/ostree', deployment])
+                if os.path.isdir(deployment_dir):
+                    files = sorted(os.listdir(deployment_dir))
+                    for f in files:
+                        kernel_file = os.sep.join([self.root_dir, 'boot/ostree', deployment, f])
+                        version_match = re.match(kernel_ostree_pattern, kernel_file)
+                        if version_match:
+                            version = version_match.group(1)
+                            return kernel_type(
+                                name=os.path.basename(
+                                    os.path.realpath(kernel_file)
+                                ),
+                                filename=kernel_file,
+                                version=version
+                            )
```

Comment:
I think we should move that code into an extra method.

Other than that I believe it does not have unwanted side effects as it only runs if ```boot/ostree``` exists which I think is a fair check into ostree specific code.

What I wondered about is if the version lookup makes sense/is useful. iirc the code walks through all deployment directories and the first one matching ```'.*/boot/ostree/.*/vmlinuz-(.*)'``` is going to be used. Is this accurate ? 
Can there be multiple different kernel versions in the deployments ?

in kiwi the kernel version lookup is not only used for bootloader code it's for example also used to construct the proper initrd file name when calling dracut and at other places. Thus fetching the information about the later running kernel of the OS is important for the build process of images that needs a kernel (typically disk images) 

I guess it's also because I don't know the inner works of ostree, but correct me when wrong, in the later running system there is still a kernel and its modules at the expected places ```/lib/modules/<kver>/...``` or the ```usr``` as I cannot imagine that ostree systems has moved everything to ```/boot/ostree/<deployment-id>/...```, correct ?

If so I also wondered why the kiwi built could not find a kernel at the usual place

Reply:
> Can there be multiple different kernel versions in the deployments ?

ostree only supports having a single kernel per deployment.

> in the later running system there is still a kernel and its modules at the expected places

Yes, on the final system, things are mounted such that you have the impression of being on a classic system. You don't see the "real root" layout with `deploy`, etc.

> If so I also wondered why the kiwi built could not find a kernel at the usual place

Kiwi is looking at the "real root" layout, not at the final layout that is setup in the initramfs by ostree for the final system, so `/lib/modules` does not exists there, only in the deployment.

Overall, it might make sense to completely skip any kernel lookup in the ostree case.